### PR TITLE
BBS+ Signatures FFI Wrapper for Python

### PR DESCRIPTION
**Pull request by karimStekelenburg**
Contains a wrapper for Python. Tests and test pipelines are included for macOS, Linux and Windows.

TODOs:
- [x] In code documentation
- [x] Replace `c_uint` types in FFI layer with more explicit unsigned integer types (see [Andrew's comment](https://github.com/mattrglobal/ffi-bbs-signatures/pull/22#discussion_r606393602))
- [x] Fix ursa lib file names in `setup.py` (see: [Andrew's comment](https://github.com/mattrglobal/ffi-bbs-signatures/pull/22#discussion_r606391064))
- [x] CI actions for releasing to PyPi

### CODE REVIEW

**Comment by karimStekelenburg** [comment]
Hi @tplooker @andrewwhitehead @swcurran! Here is my PR for the Python wrapper.
I was wondering what the right approach is for releasing it to PyPi, we were thinking the Hyperledger account might be the best place for it. Any thoughts on this?

Thanks!

**Comment by tmarkovski** [comment]
> I was wondering what the right approach is for releasing it to PyPi, we were thinking the Hyperledger account might be the best place for it. Any thoughts on this?

Same question for the .NET wrapper. Would be great to add Nuget publishing. We're currently doing it from a fork.

**Comment by swcurran** [comment]
Love to see this reviewed, accepted if appropriate, and a way to get the artifacts published.  @tplooker @kdenhartog @mikelodder7 

Thanks!

File: wrappers/python/setup.py
Comment (andrewwhitehead) [inline wrappers/python/setup.py:32]: I think this should be `bbs.dll`, `libbbs.dylib`, `libbbs.so`

File: wrappers/python/ursa_bbs_signatures/_ffi/bindings/bbs_blind_commitment.py
Comment (andrewwhitehead) [inline wrappers/python/ursa_bbs_signatures/_ffi/bindings/bbs_blind_commitment.py:2]: I think it might be good to be more explicit about the integer types, ie. `c_uint32` and `c_uint64` to avoid issues on different platforms.

File: wrappers/python/setup.py
Comment (karimStekelenburg) [reply wrappers/python/setup.py:32]: Good catch, thanks!

File: wrappers/python/ursa_bbs_signatures/_ffi/bindings/bbs_blind_commitment.py
Comment (karimStekelenburg) [reply wrappers/python/ursa_bbs_signatures/_ffi/bindings/bbs_blind_commitment.py:2]: Agreed! I wasn't sure about the integer types at first and therefore chose to use `c_uint` for the time being. I think I did so for all binding functions, so I'll run through all of them and make them more explicit. Thanks! 👌

**Comment by tplooker** [comment]
Thanks @karimStekelenburg this looks great!

> I was wondering what the right approach is for releasing it to PyPi, we were thinking the Hyperledger account might be the best place for it. Any thoughts on this?

I'm currently reviewing a couple of options, we are happy to push out the management of this wrapper to somewhere like hyperledger, if they want to take over maintenance?

**Comment by karimStekelenburg** [comment]
Thanks @tplooker! I'll make sure to finalize the remaining tasks in the coming days. 

If my understanding is correct I believe @andrewwhitehead is one of the maintainers of the Hyperledger PyPi account. @andrewwhitehead if my assumption is correct, would you be willing to push it to the registry? I'll gladly volunteer to help maintaining the code.

**Comment by andrewwhitehead** [comment]
I'm not a maintainer, Ry Jones usually manages that. Generally you would publish a pypi package first and add Hyperledger as an owner later, and then add the authentication token as a secret on a hyperledger repo to enable workflow-based publishing.

**Comment by karimStekelenburg** [comment]
Thanks @andrewwhitehead! I've made the changes and I think it’s ready to be merged. 

I have added a CI action that pushes the package to PyPi and have published an initial version: [ursa-bbs-signatures · PyPI](https://pypi.org/project/ursa-bbs-signatures/). The publish action will only run when triggered manually, or when a release is created, so this shouldn’t be a blocker for the merge.

I’ll shoot a message to Ry Jones about how to go about the ownership of the package with regards to publishing secrets etc.

**Comment by ryjones** [comment]
Once this is merged into a hyperledger repo, we can talk about how to publish. it is fairly straightforward.

**Review by tplooker — APPROVED** [review]

**Comment by tplooker** [comment]
PR LGTM, merging as is, then we can co-ordinate next steps around the wrappers home

**Pull request merged**
